### PR TITLE
feat: Adds AWS KMS signing.

### PR DESCRIPTION
**Pull request by ianhundere**
### Description of the changes being introduced by the pull request:
This pull request adds AWS signer implementation. ~~though currently the func that gets the scheme and keytype just provides a default signing algorithm for each respective key spec specifically for RSA keys.~~

~~I added logic to the sign func that parses the correct signing algorithm which is a bit messy, but I had trouble trying to pull the scheme in from the test elsewhere and didn't want to muddy the way things were done compared to the gcp/azure implementations.~~

I also redacted the necessary information in the test, but I also see that the azure test isn't currently being used in pipelines.

### Please verify and check that the pull request fulfils the following requirements:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by ianhundere** [comment]
Ah, I haven't had a chance to implement this but I'll try the following:
```
        signing_algos = ["RSASSA_PSS_SHA_256", "RSASSA_PSS_SHA_384", "RSASSA_PSS_SHA_512", "RSASSA_PKCS1_V1_5_SHA_256", "RSASSA_PKCS1_V1_5_SHA_384", "RSASSA_PKCS1_V1_5_SHA_512"]
        keytype_and_scheme = {
            "ECC_NIST_P256": [
                ("ecdsa", "ECDSA_SHA_256"),
            ],
            "ECC_NIST_P384": [
                ("ecdsa", "ECDSA_SHA_384"),
            ],
            "RSA_2048": ("rsa", signing_algos),
            "RSA_3072": ("rsa", signing_algos),
            "RSA_4096": ("rsa", signing_algos),
        }
```
I'll then extract the correct scheme by matching it against the scheme in the test input data and then pass `keytype` / `scheme` to the `sign` func. Should be able to do that tomorrow morning.

**Comment by lukpueh** [comment]
Very cool! Thanks for the contribution, @ianhundere! From a quick glance the code looks great. May I ask you to fix ci failures?
- Windows builds fail, because we seem to drop right below the minimum coverage there. As a quick fix, I suggest to just further lower the coverage constraint in tox.ini. We can fix it properly later with https://github.com/secure-systems-lab/securesystemslib/issues/208
- The purepy build fails, because `cryptography` is also an optional dependency and thus should go into the try block like `boto`
- Lint build fails, because code does not conform to the black style. You can test all linters locally with `tox -e lint`.

> I added logic to the sign func that parses the correct signing algorithm which is a bit messy, but I had trouble trying to pull the scheme in from the test elsewhere and didn't want to muddy the way things were done compared to the gcp/azure implementatios.

Your code looks fine for a prototype. We can polish later, e.g. with https://github.com/secure-systems-lab/securesystemslib/issues/594 and https://github.com/secure-systems-lab/securesystemslib/issues/593.

> I also redacted the necessary information in the test, but I also see that the azure test isn't currently being used in pipelines.

Did you test this somewhere? Would you mind providing some details about how to setup AWS KMS to use AWSSigner?

**Comment by ianhundere** [comment]
Appreciate the feedback, @lukpueh. Just pushed a commit that ensures the keytype and scheme passed in import_ are consistent with the actual signing algorithm, tests were passing / correct signing algo was being passed, but the `scheme` value wasn't always consistent with the algo passed to aws kms. 

I'll implement your feedback over the weekend, cheers / thanks. 🍻

**Comment by ianhundere** [comment]
Okay, I managed to resolve most all `tox -e lint` errors except for:
`error: Unpacking a string is disallowed  [misc]`

though seems there is some disagreement on this and that there'll be an option to bypass it soon as per:
- https://github.com/python/mypy/pull/15511

Anyhoo, this should be all done.

**Comment by ianhundere** [comment]
@lukpueh Okay, all `tox -e lint` should now be resolved.

**Comment by ianhundere** [comment]
@lukpueh added one last change to cleanup `_get_keytype_and_scheme` by adding a new func, `_get_aws_signing_algo`, and updated some docstrings, too. should be all good now.

```
Success: no issues found in 14 source files
  lint: OK (7.78=setup[0.05]+cmd[0.71,0.30,5.79,0.49,0.46] seconds)
  congratulations :) (7.84 seconds)
  ```

**Comment by lukpueh** [comment]
Thanks again for the very valuable contribution! The AWS KMS interaction looks great. 💯 

I did have a hard time, though, to follow your signing scheme detection code. It looks like you are putting a lot of effort into making the `keytypes_and_schemes` map bidirectional? ... while taking into account the signing algorithms that are supported by the KMS?

I have two comments about that:

1. The scheme in the public key object that you pass to the aws signer constructor should be the single source of truth.

    This means, you can ask the KMS for available signing algorithms when you import the public key, and just pick any one, e.g the first as you do now. But when you create or use the signer, you should not ask the KMS again, but instead use the signature algorithm appropriate for the public key signing scheme.

2. There are many ways to make `keytypes_and_schemes` bidirectional. Given the small size of the map, I think a reasonable and very readable option is to just duplicate it. E.g.:
    ```python
    def _get_keytype_and_scheme(aws_signing_algorithm: str) -> Tuple[str, str]:
        keytypes_and_schemes = {
            "ECDSA_SHA_256": ("ecdsa", "ecdsa-sha2-nistp256"),
            ...
            }
        return keytypes_and_schemes[aws_signing_algorithm]

    def _get_aws_signing_algorithm(scheme: str) -> str:
        aws_signing_algorithms = {
            "ecdsa-sha2-nistp256": "ECDSA_SHA_256",
            ...
            }
        return aws_signing_algorithms[scheme]
    ```

Apologies, if I am missing something. Also, happy to discuss this in person.

**Review by lukpueh — COMMENTED** [review]
Here's a couple of docs and style -related comments. Definitely no blockers, but if you have time, it would be nice to fix them.

**Comment by ianhundere** [comment]
@lukpueh okay, all suggestions implemented, much prettier to look at now. 😅 

thanks again for the feedback. 🙂

**Review by lukpueh — COMMENTED** [review]
Thanks for the quick updates, this looks great! 

I like that you just let the user decide, which scheme to use. We can make enhancements later, like cross-check with the schemes that are actually supported, or pick a reasonable default from the supported schemes, etc.

I left a few final comments. Otherwise, I think this is ready to merge, and we can take a stab at testing in a follow-up PR (#612).

File: securesystemslib/signer/_aws_signer.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+        keyid = cls._get_keyid(keytype, scheme, keyval)
+        public_key = SSlibKey(keyid, keytype, scheme, keyval)
+        return f"{cls.SCHEME}:{aws_key_id}", public_key
+
+    @staticmethod
+    def _get_keytype_and_scheme(
+        scheme: str,
+    ) -> Tuple[str, str]:
+        """Returns the Secure Systems Library key type and scheme for the AWS KMS
+        key type and signing algorithm
```
[the remark this answers is not in the document]
Comment (ianhundere) [reply securesystemslib/signer/_aws_signer.py:142]: ah, that makes a lot of sense. i was trying so hard to maintain the pattern(s) used before it caused me to overthink everything. again, thanks for the feedback. just pushed your suggestions.

**Review by lukpueh — APPROVED** [review]
Thanks!

**Pull request merged**